src/libumockdev-preload.c

```diff
@@ -1474,6 +1474,28 @@ int fstatfs ## suffix(int fd, struct statfs ## suffix *buf)	\
 WRAP_FSTATFS();
 WRAP_FSTATFS(64);
 
+#define WRAP_STATFS(suffix) \
+int statfs ## suffix(const char *path, struct statfs ## suffix *buf) {	\
+    libc_func(statfs ## suffix, int, const char*, struct statfs ## suffix *buf); \
+    int r;								\
+    TRAP_PATH_LOCK;							\
+    const char *p = trap_path(path);					\
+    if (p == NULL || p == path) {					\
+	r = _statfs ## suffix(path, buf);				\
+	TRAP_PATH_UNLOCK;						\
+	return r;							\
+    } \
+    DBG(DBG_PATH, "testbed wrapped statfs" #suffix "(%s) -> %s\n", path, p); \
+    r = _statfs ## suffix(p, buf);					\
+    TRAP_PATH_UNLOCK;							\
+    if (r == 0 && (strcmp(path, "/sys") == 0 || strncmp(path, "/sys/", 5) == 0))	\
```

Comment:
I am starting to think that a helper function like `is_dir_or_contained` or so would be useful. The strcmp/strncmp is not entirely obvious to me when reading it and it is needed in multiple places. It could be inlined so that the compiler can optimize away the `strlen()` call to be static.

Reply:
Good idea, done. I rely on gcc to figure out the optimization :grinning: